sql: fix bug where bad mutation job state could block dropping tables

thoszhang opened the pull request:
Previously, while dropping a table, we would mark all the jobs
associated with mutations on the table as `succeeded`, under the
assumption that they were running. The job registry API prohibits this
when the jobs are not `running` (or `pending`), so if a mutation was
stuck on the table descriptor with a failed or nonexistent job, dropping
the table would fail.

This PR fixes the bug by checking the job state before attempting to
update the job. It also fixes a related failure to drop a table caused
by a valid mutation job not being in a `running` state.

Fixes #57597.

Release note (bug fix): Fixed a bug where prior schema changes on a
table that failed and could not be fully reverted could prevent the
table from being dropped.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57836)
<!-- Reviewable:end -->

lucy-zhang commented:
Ultimately I went for what seemed to be the simplest fix, which is to keep marking these jobs as succeeded whenever possible. I also realized that this was also broken if we have `reverting` mutation jobs. That could use a separate test.

ajwerner reviewed (commented):
> Reviewed 1 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57836)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @lucy-zhang)

___
*[pkg/sql/drop_table.go, line 470 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-MOZXGig8HCB68Net3fL:-MOZXGig8HCB68Net3fM:b-jg4r9z) ([raw file](https://github.com/cockroachdb/cockroach/blob/429087584fa46d07c8d2935843070f6a7dcf7439/pkg/sql/drop_table.go#L470)):*
> ```Go
> 		// a failed state. Such jobs could have already been GCed from the jobs
> 		// table by the time this code runs.
> 		row, err := p.ExtendedEvalContext().InternalExecutor.(*InternalExecutor).QueryRowEx(
> ```

Is this really better than loading the job and reading its status using the jobs APIs? 


<!-- Sent from Reviewable.io -->

Commit:
sql: fix bug where bad mutation job state could block dropping tables

Previously, while dropping a table, we would mark all the jobs
associated with mutations on the table as `succeeded`, under the
assumption that they were running. The job registry API prohibits this
when the jobs are not `running` (or `pending`), so if a mutation was
stuck on the table descriptor with a failed or nonexistent job, dropping
the table would fail.

This PR fixes the bug by checking the job state before attempting to
update the job. It also fixes a related failure to drop a table caused
by a valid mutation job not being in a `running` state.

Release note (bug fix): Fixed a bug where prior schema changes on a
table that failed and could not be fully reverted could prevent the
table from being dropped.

thoszhang reviewed (commented):
Also added a test for when we're dropping a table with mutation jobs in the reverting state.

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-:-MP81VKu08sGg4UrFnAc:bsy9vo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/sql/drop_table.go, line 470 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-MOZXGig8HCB68Net3fL:-MP81H6p6JP4EJMoT38Y:btz58uz) ([raw file](https://github.com/cockroachdb/cockroach/blob/429087584fa46d07c8d2935843070f6a7dcf7439/pkg/sql/drop_table.go#L470)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Is this really better than loading the job and reading its status using the jobs APIs? 
</blockquote></details>

That's a fair point. I updated it to use `UpdateJobWithTxn`. I also decided I didn't want to overwrite the original error message in the case of jobs that need to be marked as failed.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Does this need to go back to 20.1 as well?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-:-MPFzWjO8yNK8eJxdjIM:b-g47p17)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/schema_changer_test.go, line 6157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-MPFzn-aDJ8lsaQ0QXBL:-MPFzn-aDJ8lsaQ0QXBM:bnuj150) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be5c5783dc39c7bf1e35cef4a77c8bdf8806a47/pkg/sql/schema_changer_test.go#L6157)):*
> ```Go
> 		// Try to create a unique index which won't be valid and will need a rollback.
> 		_, err = sqlDB.Exec(`CREATE UNIQUE INDEX i ON t.test(v);`)
> 		require.Regexp(t, "violates unique constraint", err)
> ```

Use `assert` in a goroutine.
___
*[pkg/sql/schema_changer_test.go, line 6187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-MPFzxbV76DsNVBTSaWq:-MPFzxbV76DsNVBTSaWr:blfkd9m) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be5c5783dc39c7bf1e35cef4a77c8bdf8806a47/pkg/sql/schema_changer_test.go#L6187)):*
> ```Go
> 	ctx := context.Background()
> 
> 	runTest := func(params base.TestServerArgs, gcJobRecord bool) {
> ```

pass in the `testing.T`


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
I added a query before calling `JobRegistry.UpdateJobWithTxn` to check that the job exists at all (which I don't think is possible through the registry API) and to not return an error in that case. Should be fixed now.

> Does this need to go back to 20.1 as well?

Probably. I'll figure that out.

> Reviewed 2 of 2 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-:-MPGC5xu1OdM0IAYpYel:bcy95jx)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @ajwerner)

___
*[pkg/sql/schema_changer_test.go, line 6157 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-MPFzn-aDJ8lsaQ0QXBL:-MPGD2oyAYnTvbsWbPvG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be5c5783dc39c7bf1e35cef4a77c8bdf8806a47/pkg/sql/schema_changer_test.go#L6157)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

Use `assert` in a goroutine.
</blockquote></details>

Done.
___
*[pkg/sql/schema_changer_test.go, line 6187 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-MPFzxbV76DsNVBTSaWq:-MPGD89e9_KD7KjTdHm4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/0be5c5783dc39c7bf1e35cef4a77c8bdf8806a47/pkg/sql/schema_changer_test.go#L6187)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

pass in the `testing.T`
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-:-MPGDdmO0UsrQt0yXvZJ:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

lucy-zhang commented:
TFTR!

bors r+

lucy-zhang commented:
I'm realizing now that this is not going to backport cleanly to 20.2 because `UpdateJobWithTxn` is only on master, and doesn't quite do what I want anyway. So I'm going to rework this slightly and just load the job itself.

bors r-

lucy-zhang commented:
RFAL

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/57836#-:-MPHBCZn2WKC3CLp6g_s:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained



<!-- Sent from Reviewable.io -->

lucy-zhang commented:
TFTR

bors r+
